### ISSUE TITLE
lora qwen3-vl in custom datasets which is VQA data,but lora's speed is too slow.

### ISSUE THREAD

**Issue by DanielSACE** [modelscope/ms-swift#7358]
Lora qwen3-vl-2b in my custom vqa datasets, lora's speed is too slow.

I'm lora qwen3-vl-2b in 2 4080's GPU, `--train_type lora --torch_dtype float16 --num_train_epochs 2 --per_device_train_batch_size 4 --gradient_accumulation_steps 2 --learning_rate 1e-4 --lora_rank 16 --lora_alpha 32 --target_modules all-linear --freeze_vit true --max_length 2560 --gradient_checkpointing true --output_dir xxx --eval_steps 50 --save_steps 50 --save_total_limit 3 --logging_steps 10 --warmup_ratio 0.05 --dataloader_num_workers 8 --dataset_num_proc 4 --logging_steps 1`

How can I do to solve this problem?

`'loss': 5.00581551, 'grad_norm': inf, 'learning_rate': 4.9e-07, 'token_acc': 0.38139535, 'epoch': 0.0, 'global_step/max_steps': '1/4112', 'percentage': '0.02%', 'elapsed_time': '8m 37s', 'remaining_time': '24d 15h 8m 26s', 'memory(GiB)': 21.54, 'train_speed(iter/s)': 0.001932`

**Comment by slin000111** [comment]
> Lora qwen3-vl-2b in my custom vqa datasets, lora's speed is too slow.
> 
> I'm lora qwen3-vl-2b in 2 4080's GPU, `--train_type lora --torch_dtype float16 --num_train_epochs 2 --per_device_train_batch_size 4 --gradient_accumulation_steps 2 --learning_rate 1e-4 --lora_rank 16 --lora_alpha 32 --target_modules all-linear --freeze_vit true --max_length 2560 --gradient_checkpointing true --output_dir xxx --eval_steps 50 --save_steps 50 --save_total_limit 3 --logging_steps 10 --warmup_ratio 0.05 --dataloader_num_workers 8 --dataset_num_proc 4 --logging_steps 1`
> 
> How can I do to solve this problem?
> 
> `'loss': 5.00581551, 'grad_norm': inf, 'learning_rate': 4.9e-07, 'token_acc': 0.38139535, 'epoch': 0.0, 'global_step/max_steps': '1/4112', 'percentage': '0.02%', 'elapsed_time': '8m 37s', 'remaining_time': '24d 15h 8m 26s', 'memory(GiB)': 21.54, 'train_speed(iter/s)': 0.001932`

The video may have a high frame rate and large pixel count. Try adjusting these two environment variables. VIDEO_MAX_TOKEN_NUM, FPS_MAX_FRAMES.

**Comment by DanielSACE** [comment]
> > Lora qwen3-vl-2b in my custom vqa datasets, lora's speed is too slow.
> > I'm lora qwen3-vl-2b in 2 4080's GPU, `--train_type lora --torch_dtype float16 --num_train_epochs 2 --per_device_train_batch_size 4 --gradient_accumulation_steps 2 --learning_rate 1e-4 --lora_rank 16 --lora_alpha 32 --target_modules all-linear --freeze_vit true --max_length 2560 --gradient_checkpointing true --output_dir xxx --eval_steps 50 --save_steps 50 --save_total_limit 3 --logging_steps 10 --warmup_ratio 0.05 --dataloader_num_workers 8 --dataset_num_proc 4 --logging_steps 1`
> > How can I do to solve this problem?
> > `'loss': 5.00581551, 'grad_norm': inf, 'learning_rate': 4.9e-07, 'token_acc': 0.38139535, 'epoch': 0.0, 'global_step/max_steps': '1/4112', 'percentage': '0.02%', 'elapsed_time': '8m 37s', 'remaining_time': '24d 15h 8m 26s', 'memory(GiB)': 21.54, 'train_speed(iter/s)': 0.001932`
> 
> The video may have a high frame rate and large pixel count. Try adjusting these two environment variables. VIDEO_MAX_TOKEN_NUM, FPS_MAX_FRAMES.

Only pictures, no any video.

**Comment by slin000111** [comment]
> > > Lora qwen3-vl-2b in my custom vqa datasets, lora's speed is too slow.
> > > I'm lora qwen3-vl-2b in 2 4080's GPU, `--train_type lora --torch_dtype float16 --num_train_epochs 2 --per_device_train_batch_size 4 --gradient_accumulation_steps 2 --learning_rate 1e-4 --lora_rank 16 --lora_alpha 32 --target_modules all-linear --freeze_vit true --max_length 2560 --gradient_checkpointing true --output_dir xxx --eval_steps 50 --save_steps 50 --save_total_limit 3 --logging_steps 10 --warmup_ratio 0.05 --dataloader_num_workers 8 --dataset_num_proc 4 --logging_steps 1`
> > > How can I do to solve this problem?
> > > `'loss': 5.00581551, 'grad_norm': inf, 'learning_rate': 4.9e-07, 'token_acc': 0.38139535, 'epoch': 0.0, 'global_step/max_steps': '1/4112', 'percentage': '0.02%', 'elapsed_time': '8m 37s', 'remaining_time': '24d 15h 8m 26s', 'memory(GiB)': 21.54, 'train_speed(iter/s)': 0.001932`
> > 
> > 
> > The video may have a high frame rate and large pixel count. Try adjusting these two environment variables. VIDEO_MAX_TOKEN_NUM, FPS_MAX_FRAMES.
> 
> Only pictures, no any video.

Compare the two solutions: device_map and deepspeed zero3.

**Comment by DanielSACE** [comment]
> > > > Lora qwen3-vl-2b in my custom vqa datasets, lora's speed is too slow.
> > > > I'm lora qwen3-vl-2b in 2 4080's GPU, `--train_type lora --torch_dtype float16 --num_train_epochs 2 --per_device_train_batch_size 4 --gradient_accumulation_steps 2 --learning_rate 1e-4 --lora_rank 16 --lora_alpha 32 --target_modules all-linear --freeze_vit true --max_length 2560 --gradient_checkpointing true --output_dir xxx --eval_steps 50 --save_steps 50 --save_total_limit 3 --logging_steps 10 --warmup_ratio 0.05 --dataloader_num_workers 8 --dataset_num_proc 4 --logging_steps 1`
> > > > How can I do to solve this problem?
> > > > `'loss': 5.00581551, 'grad_norm': inf, 'learning_rate': 4.9e-07, 'token_acc': 0.38139535, 'epoch': 0.0, 'global_step/max_steps': '1/4112', 'percentage': '0.02%', 'elapsed_time': '8m 37s', 'remaining_time': '24d 15h 8m 26s', 'memory(GiB)': 21.54, 'train_speed(iter/s)': 0.001932`
> > > 
> > > 
> > > The video may have a high frame rate and large pixel count. Try adjusting these two environment variables. VIDEO_MAX_TOKEN_NUM, FPS_MAX_FRAMES.
> > 
> > 
> > Only pictures, no any video.
> 
> Compare the two solutions: device_map and deepspeed zero3.

deepspeed zero3 is in use, it's also slow. How do you said device_map?

**Comment by dajuanmaoa** [comment]
+1，too slow
I'm lora qwen3-vl-2b in 2 4090's GPU

```
{'loss': 5e-07, 'grad_norm': 0.00010695, 'learning_rate': 1e-06, 'completions/mean_length': 107.9375, 'completions/min_length': 73.0, 'completions/max_length': 157.0, 'completions/clipped_ratio': 0.0, 'reward': 4.0, 'reward_std': 0.0, 'frac_reward_zero_std': 1.0, 'rewards/FormatFunc/mean': 3.0, 'rewards/FormatFunc/std': 0.0, 'rewards/ConclusionFormat/mean': 1.0, 'rewards/ConclusionFormat/std': 0.0, 'kl': 0.00047263, 'clip_ratio/low_mean': 0.0, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/high_max': 0.0, 'clip_ratio/region_mean': 0.0, 'epoch': 0.03, 'global_step/max_steps': '8/240', 'percentage': '3.33%', 'elapsed_time': '50m 53s', 'remaining_time': '1d 0h 35m 45s', 'memory(GiB)': 21.26, 'train_speed(iter/s)': 0.00262}
Train:   3%|█▉                                                        | 8/240 [50:53<25:13:45, 391.49s/it]
```
```
--per_device_train_batch_size 4 --dataloader_num_workers 1 
```

**Comment by dajuanmaoa** [comment]
我找到了这个
```
Qwen3-VL最佳实践[](https://swift.readthedocs.io/zh-cn/latest/BestPractices/Qwen3-VL-Best-Practice.html#qwen3-vl)
环境准备[](https://swift.readthedocs.io/zh-cn/latest/BestPractices/Qwen3-VL-Best-Practice.html#id1)
在开始推理和训练之前，请确保您的环境已准备就绪。

pip install "transformers>=4.57" "qwen_vl_utils>=0.0.14"

pip install "ms-swift>=4.0"
# pip install "vllm>=0.11.0"  # 若使用vllm推理后端进行推理
关于训练缓慢：使用torch2.9会遇到训练（conv3d算子）缓慢的问题，请使用torch2.8尝试，参考[这个issue](https://github.com/pytorch/pytorch/issues/166122)。在 ms-swift>=3.11.2，你可以通过设置SWIFT_PATCH_CONV3D=1规避该问题，具体查看[这个issue](https://github.com/modelscope/ms-swift/issues/7108)。

关于视频数据训练卡住：使用decord后端读取视频可能导致卡住问题，参考[这个issue](https://github.com/dmlc/decord/issues/269)。你可以使用torchcodec后端，具体参考[qwen_vl_utils](https://github.com/QwenLM/Qwen3-VL/blob/50068df2334f309979ff05d75f1078c8309c63ed/qwen-vl-utils/src/qwen_vl_utils/vision_process.py#L390-L400)库。
```
添加上SWIFT_PATCH_CONV3D=1之后，速度快多了
```
CUDA_VISIBLE_DEVICES=1,2 \
SWIFT_PATCH_CONV3D=1 \
NPROC_PER_NODE=2 \
swift rlhf \
    --rlhf_type grpo \
```

**Comment by DanielSACE** [comment]
> 我找到了这个
> 
> ```
> Qwen3-VL最佳实践[](https://swift.readthedocs.io/zh-cn/latest/BestPractices/Qwen3-VL-Best-Practice.html#qwen3-vl)
> 环境准备[](https://swift.readthedocs.io/zh-cn/latest/BestPractices/Qwen3-VL-Best-Practice.html#id1)
> 在开始推理和训练之前，请确保您的环境已准备就绪。
> 
> pip install "transformers>=4.57" "qwen_vl_utils>=0.0.14"
> 
> pip install "ms-swift>=4.0"
> # pip install "vllm>=0.11.0"  # 若使用vllm推理后端进行推理
> 关于训练缓慢：使用torch2.9会遇到训练（conv3d算子）缓慢的问题，请使用torch2.8尝试，参考[这个issue](https://github.com/pytorch/pytorch/issues/166122)。在 ms-swift>=3.11.2，你可以通过设置SWIFT_PATCH_CONV3D=1规避该问题，具体查看[这个issue](https://github.com/modelscope/ms-swift/issues/7108)。
> 
> 关于视频数据训练卡住：使用decord后端读取视频可能导致卡住问题，参考[这个issue](https://github.com/dmlc/decord/issues/269)。你可以使用torchcodec后端，具体参考[qwen_vl_utils](https://github.com/QwenLM/Qwen3-VL/blob/50068df2334f309979ff05d75f1078c8309c63ed/qwen-vl-utils/src/qwen_vl_utils/vision_process.py#L390-L400)库。
> ```
> 
> 添加上SWIFT_PATCH_CONV3D=1之后，速度快多了
> 
> ```
> CUDA_VISIBLE_DEVICES=1,2 \
> SWIFT_PATCH_CONV3D=1 \
> NPROC_PER_NODE=2 \
> swift rlhf \
>     --rlhf_type grpo \
> ```

you're right, your add the line can solve this slow problem.
